src/qr/dclogin_scheme.rs

```diff
@@ -0,0 +1,402 @@
+        if let Qr::Login { options, .. } = result {
+            assert_eq!(options, LoginOptions::UnsuportedVersion(2));
+        } else {
+            bail!("wrong type");
+        }
+        let result = decode_login("dclogin:email@host.tld/?p=123456&v=5")?;
+        if let Qr::Login { options, .. } = result {
+            assert_eq!(options, LoginOptions::UnsuportedVersion(5));
+        } else {
+            bail!("wrong type");
+        }
+        Ok(())
+    }
+
+    #[test]
+    fn all_advanced_options() -> anyhow::Result<()> {
+        let result = decode_login(
+            "dclogin:email@host.tld?p=secret&v=1&ih=imap.host.tld&ip=4000&iu=max&ipw=87654&is=ssl&ic=1&sh=mail.host.tld&sp=3000&su=max@host.tld&spw=3242HS&ss=plain&sc=3",
+        )?;
+        if let Qr::Login { address, options } = result {
+            assert_eq!(address, "email@host.tld".to_owned());
+            assert_eq!(
+                options,
+                LoginOptions::V1 {
+                    mail_pw: "secret".to_owned(),
+                    imap_host: Some("imap.host.tld".to_owned()),
+                    imap_port: Some(4000),
+                    imap_username: Some("max".to_owned()),
+                    imap_password: Some("87654".to_owned()),
+                    imap_security: Some(Socket::Ssl),
+                    imap_certificate_checks: Some(CertificateChecks::Strict),
+                    smtp_host: Some("mail.host.tld".to_owned()),
+                    smtp_port: Some(3000),
+                    smtp_username: Some("max@host.tld".to_owned()),
+                    smtp_password: Some("3242HS".to_owned()),
+                    smtp_security: Some(Socket::Plain),
+                    smtp_certificate_checks: Some(CertificateChecks::AcceptInvalidCertificates),
+                }
+            );
+        } else {
+            bail!("wrong type")
+        }
+        Ok(())
+    }
+
+    #[test]
+    fn uri_encoded_password() -> anyhow::Result<()> {
+        let result = decode_login(
+            "dclogin:email@host.tld?p=%7BDaehFl%3B%22as%40%21fhdodn5%24234%22%7B%7Dfg&v=1",
+        )?;
+        if let Qr::Login { address, options } = result {
+            assert_eq!(address, "email@host.tld".to_owned());
+            assert_eq!(
+                options,
+                login_options_just_pw!("{DaehFl;\"as@!fhdodn5$234\"{}fg".to_owned())
+            );
+        } else {
+            bail!("wrong type")
+        }
+        Ok(())
+    }
+
+    #[test]
+    fn email_with_plus_extension() -> anyhow::Result<()> {
+        let result = decode_login("dclogin:usename+extension@host?p=1234&v=1")?;
+        if let Qr::Login { address, options } = result {
+            assert_eq!(address, "usename+extension@host".to_owned());
+            assert_eq!(options, login_options_just_pw!("1234".to_owned()));
+        } else {
+            bail!("wrong type")
+        }
+        Ok(())
+    }
+
+    // idea: should invalid uri encoding result in error?
```

Comment:
Sure, what else?

Reply:
this is one of these todo points I was to lazy to tackle or did not immediately know how to or did not know if its worth the effort. anyways I don't really care, if you think such a test could be useful feel free to try making one :shrug:

Reply:
so, either create a test or remove the comment. i do not think someone else will pick this task up by a random comment, that tends to be misunderstood months later.